File: frontend/src/types/communication.ts
export enum CommunicationChannel {
  EMAIL = 'email',
  SMS = 'sms',
  VOICE = 'voice',
}

export enum CommunicationIntent {
  BOOK_APPOINTMENT = 'book_appointment',
  CANCEL_APPOINTMENT = 'cancel_appointment',
  REQUEST_AVAILABILITY = 'request_availability',
  PAYMENT_REQUEST = 'payment_request',
  PAYMENT_QUESTION = 'payment_question',
  VERIFY_COVERAGE = 'verify_coverage',
  INSURANCE_QUESTION = 'insurance_question',
  LAB_RESULTS = 'lab_results',
  URGENT = 'urgent',
  GENERAL = 'general',
}

export enum MessageCategory {
  APPOINTMENT = 'appointment',
  PAYMENT = 'payment',
  INSURANCE = 'insurance',
  LAB_RESULTS = 'lab_results',
  GENERAL = 'general',
  URGENT = 'urgent',
}

export interface CommunicationMessage {
  patient_id: number;
  subject?: string;
  body: string;
  attachments?: string[];
  metadata?: Record<string, any>;
}

export interface CommunicationLog {
  id: number;
  patient_id: number;
  channel: CommunicationChannel;
  message_type: MessageCategory;
  subject?: string;
  body: string;
  status: string;
  intent?: CommunicationIntent;
  sent_at: string;
  delivered_at?: string;
  read_at?: string;
  response_received_at?: string;
  metadata?: Record<string, any>;
  error_message?: string;
}

export interface CommunicationPreference {
  id: number;
  patient_id: number;
  preferred_channel: CommunicationChannel;
  allow_sms: boolean;
  allow_voice: boolean;
  allow_email: boolean;
  allow_urgent_calls: boolean;
  allow_sensitive_emails: boolean;
  sms_consent_date?: string;
  voice_consent_date?: string;
  email_consent_date?: string;
  created_at: string;
  updated_at: string;
}

export interface CommunicationAnalytics {
  total_messages: number;
  by_channel: Record<CommunicationChannel, number>;
  by_category: Record<MessageCategory, number>;
  by_intent: Record<CommunicationIntent, number>;
  average_response_time: number;
  success_rate: number;
  escalation_count: number;
  most_effective_channel: CommunicationChannel;
  busiest_hours: number[];
  busiest_days: string[];
} 

File: backend/api/schemas/communication.py
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum

class CommunicationChannel(str, Enum):
    EMAIL = "email"
    SMS = "sms"
    VOICE = "voice"

class CommunicationIntent(str, Enum):
    BOOK_APPOINTMENT = "book_appointment"
    CANCEL_APPOINTMENT = "cancel_appointment"
    REQUEST_AVAILABILITY = "request_availability"
    PAYMENT_REQUEST = "payment_request"
    PAYMENT_QUESTION = "payment_question"
    VERIFY_COVERAGE = "verify_coverage"
    INSURANCE_QUESTION = "insurance_question"
    LAB_RESULTS = "lab_results"
    URGENT = "urgent"
    GENERAL = "general"

class MessageCategory(str, Enum):
    APPOINTMENT = "appointment"
    PAYMENT = "payment"
    INSURANCE = "insurance"
    LAB_RESULTS = "lab_results"
    GENERAL = "general"
    URGENT = "urgent"

class CommunicationMessage(BaseModel):
    patient_id: int
    subject: Optional[str] = None
    body: str
    attachments: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None

class CommunicationLogBase(BaseModel):
    patient_id: int
    channel: CommunicationChannel
    message_type: MessageCategory
    subject: Optional[str] = None
    body: str
    status: str
    intent: Optional[CommunicationIntent] = None
    metadata: Optional[Dict[str, Any]] = None

class CommunicationLogCreate(CommunicationLogBase):
    pass

class CommunicationLog(CommunicationLogBase):
    id: int
    sent_at: datetime
    delivered_at: Optional[datetime] = None
    read_at: Optional[datetime] = None
    response_received_at: Optional[datetime] = None
    error_message: Optional[str] = None

    class Config:
        orm_mode = True

class CommunicationPreferenceBase(BaseModel):
    patient_id: int
    preferred_channel: CommunicationChannel
    allow_sms: bool = True
    allow_voice: bool = True
    allow_email: bool = True
    allow_urgent_calls: bool = True
    allow_sensitive_emails: bool = True

class CommunicationPreferenceCreate(CommunicationPreferenceBase):
    pass

class CommunicationPreference(CommunicationPreferenceBase):
    id: int
    sms_consent_date: Optional[datetime] = None
    voice_consent_date: Optional[datetime] = None
    email_consent_date: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True

class CommunicationEscalationBase(BaseModel):
    communication_id: int
    original_channel: CommunicationChannel
    escalated_to: CommunicationChannel
    escalation_reason: str
    scheduled_for: datetime

class CommunicationEscalationCreate(CommunicationEscalationBase):
    pass

class CommunicationEscalation(CommunicationEscalationBase):
    id: int
    executed_at: Optional[datetime] = None
    status: str
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True

class CommunicationAnalytics(BaseModel):
    total_messages: int
    by_channel: Dict[CommunicationChannel, int]
    by_category: Dict[MessageCategory, int]
    by_intent: Dict[CommunicationIntent, int]
    average_response_time: float
    success_rate: float
    escalation_count: int
    most_effective_channel: CommunicationChannel
    busiest_hours: List[int]
    busiest_days: List[str] 